data-access-layer/bosh/BoshDirectorClient.js

```diff
@@ -542,16 +543,76 @@ class BoshDirectorClient extends HttpClient {
       if (this.deploymentIpsCache[deploymentName] !== undefined) {
         return this.deploymentIpsCache[deploymentName];
       } else {
-        return this
-          .getDeploymentInstances(deploymentName)
-          .reduce((ipList, instance) => ipList.concat(instance.ips), [])
-          .tap(response => {
-            logger.info(`Cached Ips for deployment - ${deploymentName} - `, response);
-            this.deploymentIpsCache[deploymentName] = response;
+        return this.getDeploymentIpsFromResource(deploymentName)
+          .then(ips => {
+            if (!_.isEmpty(ips)) {
+              logger.info(`[getDeploymentIps] Cached Ips for deployment - ${deploymentName} - `, ips);
+              this.deploymentIpsCache[deploymentName] = ips;
+              return ips;
+            } else {
+              return this.getDeploymentIpsFromDirector(deploymentName)
+                .then(ips => {
+                  logger.info(`[getDeploymentIps] Caching IPs on ApiServer and locally`);
+                  this.deploymentIpsCache[deploymentName] = ips;
+                  return this.putDeploymentIpsInResource(deploymentName, ips)
+                    .then(() => ips);
+                });
+            }
           });
       }
     });
+  }
 
+  getDeploymentIpsFromResource(deploymentName) {
+    logger.info(`[getDeploymentIps] making request to ApiServer for deployment - ${deploymentName}`);
+    let resourceId = utils.parseServiceInstanceIdFromDeployment(deploymentName);
+    if (deploymentName === resourceId) {
+      //don't contact to ApiServer if not a service-fabrik deployment
+      return Promise.resolve();
```

Comment:
shouldnt this instead throw an error? Here you are supressing it as it might lead to some other failure in caller. Better to throw exception.

Reply:
Just change the comment to include OOB deployment scenario.